astropy/modeling/fitting.py

```diff
@@ -610,35 +624,132 @@ def __call__(self, model, x, y, z=None, weights=None, **kwargs):
+        # then passed to the fitter, which is the historical behaviour and
+        # works even for fitters that don't understand masked arrays. For model
+        # sets, the fitter must be able to filter masked data internally,
+        # because fitters require a single set of x/y co-ordinates whereas the
+        # eliminated points can vary between models. To avoid this limitation,
+        # we could fall back to looping over individual model fits, but it
+        # would likely be fiddly and involve even more overhead (and the
+        # non-linear fitters don't work with model sets anyway, as of writing).
+
+        if len(model) == 1:
+            model_set_axis = None
+        else:
+            if not hasattr(self.fitter, 'supports_masked_input') or \
+               self.fitter.supports_masked_input is not True:
+                raise ValueError("{0} cannot fit model sets with masked "
+                                 "values".format(type(self.fitter).__name__))
+
+            # The intention appears to be that fitters use their input model's
+            # model_set_axis to determine how their input data are stacked:
+            model_set_axis = model.model_set_axis
+
+        # Construct input co-ordinate tuples for fitters & models that are
+        # appropriate for the dimensionality being fitted:
         if z is None:
-            filtered_data = y
-            for n in range(self.niter):
-                filtered_data = self.outlier_func(filtered_data - fitted_model(x),
-                                                  **self.outlier_kwargs)
-                filtered_data += fitted_model(x)
-                if weights is not None:
-                    filtered_weights = weights[~filtered_data.mask]
-                fitted_model = self.fitter(fitted_model,
-                               x[~filtered_data.mask],
-                               filtered_data.data[~filtered_data.mask],
-                               weights=filtered_weights,
-                               **kwargs)
+            coords = x,
+            data = y
         else:
-            filtered_data = z
-            for n in range(self.niter):
-                filtered_data = self.outlier_func(filtered_data - fitted_model(x, y),
-                                                  **self.outlier_kwargs)
-                filtered_data += fitted_model(x, y)
+            coords = x, y
+            data = z
+
+        # For model sets, first try passing the numpy-standard "axis" parameter
+        # to the outlier function, to make it treat each model separately. If
+        # this fails, fall back to looping over the model set below.
+        if model_set_axis is not None:
+            if 'axis' not in self.outlier_kwargs:  # allow user override
+                # This also works for False (like model instantiation):
+                self.outlier_kwargs['axis'] = tuple(
+                    n for n in range(data.ndim) if n != model_set_axis
```

Comment:
Can `model_set_axis` be stored in the model as a negative value? I seem to remember seeing that somewhere since I wrote this. Looks like I haven't allowed for negative values here (by converting them when setting `model_set_axis` above)...

Reply:
Yes, it can be negative.

Reply:
I see that axis is passed to numpy functions and they already work with negative axis. Is this a problem then?

Reply:
I think the problem is going to be `if n != model_set_axis`, causing the tuple to include all the axes and produce a single standard deviation over all the models.

Reply:
I've fixed this in my copy, though instantiating all the models I've tried with `model_set_axis=-1` produces an error.

Reply:
What kind of error? 
The parameters have to have proper dimension for a negative axis. But if it's some other error...

Reply:
Oh, it just doesn't work if I don't specify parameter values explicitly...

Reply:
I suppose because it has nothing to infer the dimensionality from (though it would be nice to say so rather than `numpy.core._internal.AxisError: axis -1 is out of bounds for array of dimension 0` :wink:.

Reply:
Ah, yes. I think this is by design. It's not just for negative axes. The fact that polynomials work is an unexpected side effect from he fact that they inherit directly from `Model`. I think.

Reply:
Oh, you mean instantiation without specifying parameter values when `n_models > 1`, don't you (as in #6096)? Right. Well, even with explicit coefficients, I seem to have trouble fitting with `model_set_axis=-1`, but that's more than I have time to troubleshoot now and I can see that the `axis` tuple now gets constructed correctly after my fairly trivial fix...

Reply:
If you have a ready example showing the problem, please file an issue.

Are you planning to push more changes here?

Reply:
Here they are. Thanks.